### PR TITLE
Add cloud functions to GCP plugin

### LINKED ISSUE #578: List and view logs for GCP Cloud Functions
Add GCP Cloud Functions as a Wash resource. Be able to list and delete them, as well as tail their logs.

API access is available via https://godoc.org/google.golang.org/api/cloudfunctions/v1, and https://cloud.google.com/functions/docs/monitoring/logging talks about accessing logs.

### PR DESCRIPTION
**Pull request by ekinanp**
Resolves #578

Signed-off-by: Enis Inan <enis.inan@puppet.com>

### CODE REVIEW

File: plugin/gcp/cloudFunction.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+type cloudFunction struct {
+	plugin.EntryBase
+	service cloudFunctionsProjectService
+}
+
+func newCloudFunction(function *cloudfunctions.CloudFunction, service cloudFunctionsProjectService) *cloudFunction {
+	cf := &cloudFunction{
+		EntryBase: plugin.NewEntry(function.Name),
```
Comment (ekinanp) [inline plugin/gcp/cloudFunction.go]: Looks like cloud function names are of the form `project/<project_id>/location/<region>/functions/<function_name>`. Given this, how should we organize them? I'm thinking `<region>::<function_name>`.

File: plugin/gcp/cloudFunction.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	cfl, err := newCloudFunctionLog(ctx, cf.service, cf.Name())
+	if err != nil {
+		return nil, err
+	}
+	return []plugin.Entry{cfl}, nil
+}
+
+func (cf *cloudFunction) Delete(ctx context.Context) (bool, error) {
+	_, err := cf.service.Projects.Locations.Functions.Delete(cf.Name()).Context(ctx).Do()
+	return false, err
```
Comment (ekinanp) [inline plugin/gcp/cloudFunction.go]: Cloud functions have some shut down stuff that needs to happen before they're deleted, which is why this is false.

File: plugin/gcp/cloudFunction.go (hunk 1 above)
Comment (MikaelSmith) [reply plugin/gcp/cloudFunction.go]: VM instances are also in a specific region/zone. As long as name uniqueness is enforced across zones I didn't put it in.

File: plugin/gcp/cloudFunction.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	plugin.EntryBase
+	service cloudFunctionsProjectService
+}
+
+func newCloudFunction(function *cloudfunctions.CloudFunction, service cloudFunctionsProjectService) *cloudFunction {
+	cf := &cloudFunction{
+		EntryBase: plugin.NewEntry(function.Name),
+		service:   service,
+	}
+	mtime, err := time.Parse(time.RFC3339, function.UpdateTime)
```
Comment (MikaelSmith) [inline plugin/gcp/cloudFunction.go]: Huh, only some services add a creation timestamp so it's in "Metadata".

Comment (ekinanp) [reply plugin/gcp/cloudFunction.go]: Yeah, it's weird.

File: plugin/gcp/cloudFunctionLog.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	afterTimestamp string
+	cfl            *cloudFunctionLog
+}
+
+func (s *cloudFunctionLogStreamer) Read(p []byte) (n int, err error) {
+	for {
+		if len(s.currentEntries) > 0 {
+			break
+		}
+		time.Sleep(2 * time.Second)
```
Comment (MikaelSmith) [inline plugin/gcp/cloudFunctionLog.go]: Why sleep for 2 seconds?

Comment (MikaelSmith) [reply plugin/gcp/cloudFunctionLog.go]: Oh, because copy will busyread.

Comment (MikaelSmith) [reply plugin/gcp/cloudFunctionLog.go]: I wish there were a better way, but I don't see it.

**Review by MikaelSmith — APPROVED** [review]

**Comment by ekinanp** [comment]
Updated to omit project + region from function name.

**Review by MikaelSmith — APPROVED** [review]

**Pull request merged**